test/activation_based/test_distributed_dtensor.py

```
@@ -0,0 +1,879 @@
+        zero_redundancy_optimizer_available=True,
+        pipelining_available=True,
+        fsdp2_available=True,
+        tensor_parallel_available=True,
+    )
+    assert recommendation.mode == "fsdp2_tp"
+    assert recommendation.memopt_level == 1
+    assert recommendation.mesh_shape == (2, 2)
+
+
+def test_recommended_pipeline_microbatches_rejects_too_small_batch():
+    with pytest.raises(ValueError, match="batch_size .* must be >= num_stages"):
```

Ruff (0.15.12)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)